docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,286 @@
+
+### Handling of Join requests
+
+Join requests in isolation currently do not have the connection stealing issue, as they already rely on the
+Function for the DevNonce check. If the current LNS is not the preferred gateway, it drops the
+message immediately.
+
+Problem
+
+- Join received from LNS1 and LNS2: LNS1 wins
+- First data message received only on LNS2, Function should inform LNS1 that it's the losing one
+  (information about the owning LNS needs to be "shared" between joins and data messages).
+
+#### Decision
+
+We decide to go with the simpler solution of closing the connection immediately after a Join and let
+the data flow re-establish it if needs to be. Disadvantage is that the first data message is more likely
+  to miss the window due to having re-establish the connection.
+  LNS also stores locally that it was the losing LNS and delays on a future re-join or future data messages.
+
+#### Alternative
+
+- Do nothing and accept that a one-off connection stealing on the first Data message can happen.
+- We do not close the connection after a Join. When the first data message arrives the Function checks
+  somehow the preferred gateway for this device when it joined: if the data message comes from the
+  same gateway it is allowed to process the message. If not, we should inform the previously owning
+  LNS to drop the connection and allow the new LNS to process the message. This was not preferred as
+  it has more complexity for unclear results.  
+
+### Main data message flow
+
+Currently, if LoRaDevice is not in the LNS cache, we search on the Function for all devices that
+have that DevAddr. Then we get all their twins which could result in a connection switch.
+
+#### Decision
+
+After loading all the twins for the LoRaDevice(s), we close the connections immediately.
+This is equivalent to how we [handle Join requests](#handling-of-join-requests)
+
+#### Alternative
+
+- Currently the DeviceGetter.GetDevice returns a list of devices that match the provided DevAddr. We
+  considered changing the Function to return a single device instead of a list and only load the
+  twin for this one on the LNS side. For this the Function should perform the Mic computation which
+  means that we would need to send the payload to the cloud. This is a deal breaker for us.
+
+#### Further processing
+
+This section uses this topology:
+
+```mermaid
+flowchart LR;
+    Device-->LBS1-->LNS1-->Function;
+    Device-->LBS2-->LNS2-->Function;
+    LNS1-->IoTHub;
+    LNS2-->IoTHub;
+```
+
+where Device sends data message A and then B.
+
+Here is a rundown of what should happen assuming both LNSs have a fresh twin (via background refresh
+or via fetching it using DeviceGetter.GetDevice). Changes are marked in **bold**:
+
+1. Device sends first data message A.
+1. We assume that LNS1 gets the message first. **LNS1 checks against
+   its in-memory dictionary DevEui -> flag** and since it has not seen this DevEui before contacts the Function.
```

Comment:
Too much detail.
```suggestion
1. We assume that LNS1 gets the message first. **LNS1 checks against
   its in-memory state** and since an owner for the connection was not elected yet, directly contacts the function without delay.
```